core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1377,7 +1381,14 @@ private PlanWithProperties arbitraryDistributeUnion(
                 // instead of "arbitraryPartition".
                 return new PlanWithProperties(node.replaceChildren(partitionedChildren));
             }
-            // Trino currently cannot execute stage that has multiple table scans, so in that case
+
+            // all children can be source partitioned and operate on one catalog
+            if (collectRemoteExchangeNodes(partitionedChildren, exchangeNode -> exchangeNode.getType() != REPLICATE).isEmpty()
```

Comment:
`collectRemoteExchangeNodes` -> `countRepartitionedRemoteExchangeNodes(partitionedChildren) == 0`
+ comment that this enforces no node partitioning which doesn't work with source table scans


`hasPartitionedConnectorSource` -> `countPartitionedConnectorSource(partitionedChildren) == 0`